modules/standard/Sets.chpl

```diff
@@ -0,0 +1,496 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+  This modules contains the implementation of the set type.
+*/
+module Sets {
+
+  pragma "no doc"
+  config param _sanityChecks = true;
+
+  //
+  // Some asserts are useful while developing, but can be turned off when the
+  // implementation is correct.
+  //
+  private inline proc _sanity(expr: bool) {
+    if _sanityChecks then
+      assert(expr);
+  }
+
+  //
+  // We can change the lock type later. Use a spinlock for now, even if it
+  // is suboptimal in cases where long critical sections have high
+  // contention (IE, lots of tasks trying to insert into the middle of this
+  // list, or any operation that is O(n)).
+  //
+  pragma "no doc"
+  type _lockType = ChapelLocks.chpl_LocalSpinlock;
+
+  //
+  // Use a wrapper class to let list methods have a const ref receiver even
+  // when `parSafe` is `true` and the list lock is used.
+  //
+  pragma "no doc"
+  class _LockWrapper {
+    var lock$ = new _lockType();
+
+    inline proc lock() {
+      lock$.lock();
+    }
+
+    inline proc unlock() {
+      lock$.unlock();
+    }
+  }
+
+  /*
+    A set is a...
+  */
+  record set {
+
+    /* The type of the elements contained in this set. */
+    type eltType;
+
+    /* If `true`, this set will perform parallel safe operations. */
+    param parSafe = false;
+
+    pragma "no doc"
+    var _size = 0;
+
+    pragma "no doc"
+    var _lock$ = if parSafe then new _LockWrapper() else none;
+
+    /*
+      Initializes an empty set containing elements of the given type.
+
+      :arg eltType: The type of the elements of this set.
+      :arg parSafe: If `true`, this set will use parallel safe operations.
+    */
+    proc init(type eltType, param parSafe=false) {
+    }
+
+    /*
+      Initializes a set containing elements that are copy initialized from
+      the elements contained in another set.
+
+      :arg other: The set to initialize from.
+      :arg parSafe: If `true`, this set will use parallel safe operations.
+    */
+    proc init=(const ref other: set(?t), param parSafe=other.parSafe) {
+    }
+
+    pragma "no doc"
+    inline proc deinit() {
+    }
+
+    pragma "no doc"
+    inline proc _enter() {
+      if parSafe then
+        _lock$.lock();
+    }
+
+    pragma "no doc"
+    inline proc _leave() {
+      if parSafe then
+        _lock$.unlock();
+    }
+
+    /*
+      Add a copy of the element `x` to this set. Does nothing if this set
+      already contains an element equal to the value of `x`.
+
+      :arg x: The element to add to this set.
+    */
+    proc add(const ref x: eltType) {
```

Comment:
Wouldn't you want this to have `in` intent just like the list append?

Reply:
It is const ref because it makes a copy internally. I can revert to using in intent when #13225 has been resolved.

Reply:
Ah wait. I haven't decided on the implementation yet, so `in` intent may well be reasonable here.

Reply:
But we're doing API review right now anyway....

Reply:
Yes we are, so I can give the method the `in` intent and change it later should it clash with the implementation.